src/main/kotlin/ee/carlrobert/codegpt/codecompletions/psi/PythonContextFinder.kt

```diff
@@ -0,0 +1,76 @@
+package ee.carlrobert.codegpt.codecompletions.psi
+
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.vfs.VirtualFile
+import com.intellij.psi.*
+import com.intellij.psi.util.findParentOfType
+import com.jetbrains.python.psi.PyClass
+import com.jetbrains.python.psi.PyFunction
+import com.jetbrains.python.psi.PyReferenceExpression
+import com.jetbrains.python.psi.resolve.PyResolveContext
+import com.jetbrains.python.psi.resolve.PyResolveUtil
+import com.jetbrains.python.psi.types.TypeEvalContext
+
+class PythonContextFinder : LanguageContextFinder {
+
+    /**
+     * Finds enclosing [PyFunction] or [PyClass] of [psiElement] and
+     * determines source code files of all used [PyReferenceExpression]s for the context.
+     */
+    override fun findContextSourceFiles(psiElement: PsiElement, editor: Editor): Set<VirtualFile> {
+        var significantElement = psiElement
+        // In Python whitespaces have semantic meaning
+        // its enclosing PyFunction is not inevitably its parent node, but its previous sibling
+        while (significantElement is PsiWhiteSpace) {
+            significantElement = significantElement.prevSibling
+        }
```

Comment:
There is probably room for refactoring/optimization here, to e.g. use [PsiTreeUtil.getPrevSiblingOfType](https://github.com/JetBrains/intellij-community/blob/master/platform/core-api/src/com/intellij/psi/util/PsiTreeUtil.java#L511)